src/solver/simulation/remix-storage/shave-peaks-by-remix-storage-gen.cpp

```diff
@@ -0,0 +1,185 @@
+                       unsigned hourOfMaxGen,
+                       const std::vector<double>& TotalGen,
+                       std::shared_ptr<IStorageForRemix> storage)
+{
+    double maxVariation = std::max(TotalGen[hourOfMaxGen] - TotalGen[hourOfMinGen], 0.);
+    double maxExchangeFromStorage = storage->maxExchange(hourOfMaxGen, hourOfMinGen);
+    return std::max(std::min(maxExchangeFromStorage, maxVariation / 2.), 0.);
+}
+
+struct Exchange
+{
+    bool valid()
+    {
+        return amount > eps && hourOfMinGen.has_value() && hourOfMaxGen.has_value();
+    }
+
+    std::optional<unsigned> hourOfMinGen;
+    std::optional<unsigned> hourOfMaxGen;
+    double amount = 0;
+};
+
+static Exchange searchForExhange(const std::set<unsigned>& validHours,
+                                 std::vector<double>& TotalGen,
+                                 std::vector<double>& UnsupE,
+                                 std::shared_ptr<IStorageForRemix> storage)
+{
+    auto totalGenProjection = [&](int h) { return TotalGen[h]; };
+
+    std::set<unsigned> validHoursForMin(validHours);
+    std::erase_if(validHoursForMin, [&](int h) { return UnsupE[h] <= eps; });
+
+    while (!validHoursForMin.empty())
+    {
+        auto hourOfMinGen = rng::min_element(validHoursForMin, {}, totalGenProjection);
+
+        std::set<unsigned> validHoursForMax(validHours);
+        double totalGenMin = TotalGen[*hourOfMinGen];
+        std::erase_if(validHoursForMax, [&](int h) { return TotalGen[h] < totalGenMin + eps; });
+
+        while (!validHoursForMax.empty())
+        {
+            auto hourOfMaxGen = rng::max_element(validHoursForMax, {}, totalGenProjection);
+            auto exchange = computeExchange(*hourOfMinGen, *hourOfMaxGen, TotalGen, storage);
+
+            if (exchange > eps)
+            {
+                return {*hourOfMinGen, *hourOfMaxGen, exchange};
+            }
+            validHoursForMax.erase(hourOfMaxGen);
+        }
+        validHoursForMin.erase(hourOfMinGen);
+    }
+    return {};
+}
+
+void update(Exchange& exchange,
+            std::shared_ptr<IStorageForRemix> storage,
+            std::vector<double>& UnsupE,
+            std::vector<double>& TotalGen)
+{
+    double amount = exchange.amount;
+    unsigned hourOfMinGen = exchange.hourOfMinGen.value();
+    unsigned hourOfMaxGen = exchange.hourOfMaxGen.value();
+
+    storage->withdrawal()[hourOfMaxGen] -= amount;
+    storage->withdrawal()[hourOfMinGen] += amount;
+    storage->update();
+
+    UnsupE[hourOfMaxGen] += amount;
+    UnsupE[hourOfMinGen] -= amount;
+
+    TotalGen[hourOfMaxGen] -= amount;
+    TotalGen[hourOfMinGen] += amount;
+}
+
+auto removeStorageFromList(std::vector<std::shared_ptr<IStorageForRemix>>::iterator storage,
+                           ListStorageForRemix& listStorage)
+{
+    auto d = std::distance(listStorage.begin(), storage);
+    listStorage.erase(storage, storage + 1);
+    return listStorage.begin() + d;
+}
```

Comment:
What does this function do ? I suppose it erases an iterator, and returns the next element. What about if the deleted element was the last ?